### PR TITLE
Extended address endpoint

### LINKED ISSUE #102: Add query params in address route
Add route to fetch `addresses` based on city, state or country

### Requirements
- [ ] Add route with query param `/address?city=xyz`
- [ ] Add route with query param `/address?state=xyz`
- [ ] Add route with query param `/address?country=xyz`

### PR DESCRIPTION
**Pull request by Samridhi-98**
# Related Issue

-   closes #102 

# Proposed Changes

- [x]  Added route with query param `/address?city=xyz`
- [x]  Added route with query param `/address?state=xyz`
- [x] Added route with query param  `/address?country=xyz`

# Checklist

-   [x] Tests
-   [x] Translations

### CODE REVIEW

**Comment by Samridhi-98** [comment]
@ShauryaAg  @paramsiddharth  Please review my PR.

File: data/addressdata.json [hunk 1]
```diff
@@ -1,72 +1,72 @@
...
             "pincode": "401202"
         },
         {
             "street": "6 Shammik chamber devaji RPD ",
-            "city": "Lucknow",
-            "state": "Uttar Pradesh",
-            "country": "India",
+            "city": "lucknow",
+            "state": "uttar pradesh",
+            "country": "india",
```
Comment (ShauryaAg) [inline data/addressdata.json:69]: Please don't convert all of this to lower case. Use [`.toLowerCase()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/toLowerCase) function while checking the condition.

File: src/api/routes/address.ts [hunk 2]
```diff
@@ -6,8 +6,35 @@ const router = express.Router();
...
 router.get("/", (req: Request, res: Response, next: NextFunction) => {
     try {
-        console.log(data);
-        res.status(200).json(data);
+        const city: any = req.query.city;
+        const state: any = req.query.state;
+        const country: any = req.query.country;
+        let result = [];
+
+        for (let i = 0; i < data.data.length; i++) {
```
Comment (ShauryaAg) [inline src/api/routes/address.ts:14]: Don't use a for loop to iterate over all of them. Use [`.filter()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/filter) method.

**Review by ShauryaAg — CHANGES_REQUESTED** [review]

**Comment by Samridhi-98** [comment]
@ShauryaAg  Done! please review

**Review by ShauryaAg — CHANGES_REQUESTED** [review]
We can also chain multiple `.filter()` methods to query based on multiple things at once.

Something like
```js
result = data.data
          .filter(condition1) // check for city
          .filter(condition2) // check for state
          .filter(condition3) // check for country
```

So that user may query like `/address?state=xyz&city=abc`

Please see if you can implement this, otherwise feel free contact us on the discord channel.

**Comment by Samridhi-98** [comment]
@ShauryaAg  Done, updated filter chaining as you suggested on discord please review.

File: src/api/routes/address.ts [hunk 3]
```diff
@@ -6,8 +6,26 @@ const router = express.Router();
...
-        console.log(data);
-        res.status(200).json(data);
+        const city: any = req.query.city;
+        const state: any = req.query.state;
+        const country: any = req.query.country;
+        if (Object.keys(req.query).length) {
+            if (!city && !state && !country)
+                res.status(404).json({ message: "Invalid query" });
+            else {
+                let result = data.data
```
Comment (ShauryaAg) [inline src/api/routes/address.ts:16]: change `let` to `const`

**Review by ShauryaAg — CHANGES_REQUESTED** [review]

File: src/api/routes/address.ts [hunk 4]
```diff
@@ -6,8 +6,26 @@ const router = express.Router();
...
+                    .filter((address: any) => address.city === (city ?? address.city))
+                    .filter((address: any) => address.state === (state ?? address.state))
+                    .filter((address: any) => address.country === (country ?? address.country));
+
+                if (result.length)
+                    res.status(200).json({ data: result });
+                else
+                    res.status(404).json({ message: "No data found" });
```
Comment (ShauryaAg) [inline src/api/routes/address.ts:24]: `res.status` should be `204` when no records are found.

**Review by ShauryaAg — CHANGES_REQUESTED** [review]

**Comment by Samridhi-98** [comment]
@ShauryaAg  Done!

**Review by paramsiddharth — APPROVED** [review]

**Comment by paramsiddharth** [comment]
@Samridhi-98 Thank you!

**Pull request merged**